### PR TITLE
Add flags to borg extract Command (Fixes #8564)

### LINKED ISSUE #8564: No dry-run flags when using borg extract
Using borg 1.4, I checked the use of "flags" when doing a dry-run for different borg commands. 

The command `borg extract --dry-run --list` does not print the flags (that are used with e.g. borg create / borg recreate
Instead, `borg extract` prints only the list of items that would be included in the extract without the --dry-run. 

This was the command, I used: 
`borg extract REPO::ARCHIVE /home/user/test Admin -e 'home/user/test/*.bak'  --dry-run --verbose --list `

In the name of consistency, dry-run flags could be added to borg extract.

### PR DESCRIPTION
**Pull request by alighazi288**
## Summary 
This pull request addresses the issue with the `borg extract` command not displaying the flags as seen in other commands eg. `borg create` and `borg recreate`.

### CODE REVIEW

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:74]: check the meaning of the flags (see our docs) and try running borg create with some excludes. borg extract should flag similarly to create.

Comment (ThomasWaldmann) [reply src/borg/archiver/extract_cmd.py:74]: also be careful with print, print outputs to stdout and borg extract also has the option to write extracted file contents to stdout - these 2 shall not be mixed up.

Comment (alighazi288) [reply src/borg/archiver/extract_cmd.py:74]: Apologies for the oversight. Trying a new fix and uploading soon for review.

**Comment by ThomasWaldmann** [comment]
BTW, don't add `coverage.*` files to the repo, these are temporary files.

**Comment by alighazi288** [comment]
Got it, I’ll ensure temporary files are omitted moving forward. Could you please share more details or context about the errors above? This would help me better understand and address the issue.

**Review by ThomasWaldmann — COMMENTED** [review]
some more feedback.

**Comment by ThomasWaldmann** [comment]
The failing `test_return_codes` test is because `borg extract` is used there and expected to finish with rc 0, but it crashes and returns rc 2.

**Comment by alighazi288** [comment]
Changed the code and now everything seems to be running as expected at my end and **borg extract** finishes with rc 0. Just wanted to ask if the errors could be because of me working on MacOS and not having pyfuse3?

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:75]: this log output is different from the other one, no status char.

do you practically try your code after making changes?

Comment (ThomasWaldmann) [reply src/borg/archiver/extract_cmd.py:75]: you also can run the unit tests locally, just invoke pytest or tox.

**Comment by ThomasWaldmann** [comment]
The errors are unrelated to macOS and not having pyfuse3 - I also work on a macbook.

If one want to test or work on the mount / FUSE code in borg, one would need macFUSE installed and use llfuse (not pyfuse3). But this is not required here.

File: src/borg/archiver/extract_cmd.py
Comment (alighazi288) [reply src/borg/archiver/extract_cmd.py:75]: @ThomasWaldmann ?

**Comment by alighazi288** [comment]
@ThomasWaldmann Please review.

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:105]: It would be better to display included and excluded items both together in the order they appear in the archive.

Comment (alighazi288) [reply src/borg/archiver/extract_cmd.py]: @ThomasWaldmann done.

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:59]: Hmm, this develops to be a bit more tricky than I initially thought it is.

If you leave away the filter here (which at first looks like a good idea to get ALL items yielded), it will preload the file content chunks of ALL yielded items (not just the ones we might extract).

That's not necessary and it is even harmful, leading to a memory leak if one does not retrieve these chunks, see comment in `unpack_many`.

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [reply src/borg/archiver/extract_cmd.py:65]: also `stripped_path` would be a better name, there is no sanitation done here.

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:78]: use dry_run.

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:72]: hmm, doesn't the original code match using the full (non-stripped) path?

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:90]: can it ever be False here?

**Review by ThomasWaldmann — CHANGES_REQUESTED** [review]
This is getting tricky.

File: src/borg/archiver/extract_cmd.py
Comment (alighazi288) [reply src/borg/archiver/extract_cmd.py:59]: When you use the filter, only items that match the filter criteria are processed which prevents us from logging any excluded items, which is why I proposed the "two pass" solution but that would display excludes either at the beginning or the end.

Comment (ThomasWaldmann) [reply src/borg/archiver/extract_cmd.py:59]: Sure. So there is no good solution yet and it is not easy due to the preloading requirements.

Comment (ThomasWaldmann) [reply src/borg/archiver/extract_cmd.py:59]: I am working right now on decoupling preloading and item iteration to help you with that.

After that is merged, you can rebase your work on these changes.

Comment (alighazi288) [reply src/borg/archiver/extract_cmd.py:59]: Thank you so much. I really appreciate it!

Comment (ThomasWaldmann) [reply src/borg/archiver/extract_cmd.py:59]: #8592 was just merged into master.

Please update your local master branch and rebase your PR branch onto that.

As you can see there, `archive.iter_items()` does not support preloading anymore.
At the places where preloading is needed, I inserted a new call that does the preloading.

Be careful: the preload call must only be done if the item chunks will be fetched later (e.g. if the item is extracted, because it is included / not excluded). The `borg extract` code fetches the content chunks even if it is in dry-run mode, because it is documented to do all steps of the extraction except writing content to disk files (so people can use it as a practical test whether the extraction would work without needing the space to do the extraction).

**Comment by ThomasWaldmann** [comment]
Can you rather rebase on master please than merge master in here?

File: src/borg/archiver/extract_cmd.py
Comment (alighazi288) [reply src/borg/archiver/extract_cmd.py]: Thank you so much once again for the clarification and your help! Working on it.

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [reply src/borg/archiver/extract_cmd.py]: also, does it have a special reason why you moved the extraction to here and not kept it at the original place? the code in 91+ extracts needed dirs and that should be done before files are extracted that have these as parent dirs.

**Comment by alighazi288** [comment]
I wanted to let you know that there appears to be a large number of changes in this pull request because I encountered some issues while attempting to rebase on master. I apologize for any confusion this may have caused.

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [reply src/borg/archiver/extract_cmd.py]: Maybe it makes sense to keep the test and start from scratch with the extraction code changes.

Now that the preloading is out of the way, it should be easy to get a list of ALL items, do the matching here, extract and "+" if it matches, "-" if it doesn't.

Comment (ThomasWaldmann) [reply src/borg/archiver/extract_cmd.py]: My comment above was written before the most recent changes, so let's close this.

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:58]: Isn't "None" the default anyway?

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:71]: Why did you remove checking output_list?

**Review by ThomasWaldmann — CHANGES_REQUESTED** [review]
That looks quite ok now, some minor things todo!

File: src/borg/archiver/extract_cmd.py
Comment (alighazi288) [reply src/borg/archiver/extract_cmd.py:71]: I thought it simplified the logging logic. In the updated code, we always log the paths of the items with a prefix indicating whether they are matched or not. If it is necessary to conditionally control the logging based on the output_list flag, I can reintroduce that check.

File: src/borg/archiver/extract_cmd.py
Comment (alighazi288) [reply src/borg/archiver/extract_cmd.py:58]: Yes, you're right. I thought passing None would make the code's behavior more explicit and clear. I can remove it to simplify the code.

File: src/borg/archiver/extract_cmd.py
Comment (alighazi288) [reply src/borg/archiver/extract_cmd.py:71]: @ThomasWaldmann Yes it is. However, I have still added the output_list as per your request. Please let me know what else is required to finally close this?

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:71]: please avoid such comments. it does not add any information that is not obvious from the next 2 lines anway.

Comment (alighazi288) [reply src/borg/archiver/extract_cmd.py:71]: Understood. Would you like me to remove it and commit again?

Comment (ThomasWaldmann) [reply src/borg/archiver/extract_cmd.py:71]: Yes, please.

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:66]: the original code only did that "if strip_components".

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:42]: as you are back to the original code here, move it 1 line up, so it disappears from the diff.

also, in case you don't do that already, I'ld suggest you read the diff, so you can find such things by yourself.

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [reply src/borg/archiver/extract_cmd.py:66]: also, you could stay closer to the original code (like: "don't fix what ain't broken") and just add the continue if item.path ends up being empty.

File: src/borg/archiver/extract_cmd.py
Comment (ThomasWaldmann) [inline src/borg/archiver/extract_cmd.py:69]: if you don't use log_prefix except in line 72, move 69 directly above that line.

**Review by ThomasWaldmann — COMMENTED** [review]
some minor nitpicks, please fix.

it's mostly about staying closer to the original src and not doing unneeded changes.

guess it can be merged after that (I will squash all the commits, so history will be cleaner). thanks for working on this!

**Comment by ThomasWaldmann** [comment]
There is still that ".coverage 2" file in your PR, you need to remove it, it is a temporary file and must not be committed.

File: src/borg/archiver/extract_cmd.py
Comment (alighazi288) [reply src/borg/archiver/extract_cmd.py]: Got it. The only issue is, that just adding continue resulted in test failures due to a `ValueError` being raised. `ValueError: path '' is not sanitized`. To resolve this, I updated the code to ensure that `item.path` is only set if the resulting path is not empty.

**Review by ThomasWaldmann — APPROVED** [review]
LGTM.

**Pull request merged**
